src/stream/src/executor/over_window/mod.rs

```diff
@@ -0,0 +1,692 @@
+                            &key.encoded_pk,
+                            &this.pk_data_types,
+                            &vec![OrderType::ascending(); this.input_pk_indices.len()],
+                        )?;
+                        let state_row_pk = (&partition_key)
+                            .chain(row::once(Some(key.order_key)))
+                            .chain(pk);
+                        // NOTE: We don't know the value of the row here, so the table must allow
+                        // inconsistent ops.
+                        this.state_table.delete(state_row_pk);
+                    }
+                }
+            }
+        }
+
+        let columns: Vec<Column> = builders
+            .into_iter()
+            .map(|b| b.finish().into())
+            .collect_vec();
+        let chunk_size = columns[0].len();
+        Ok(if chunk_size > 0 {
+            Some(StreamChunk::new(
+                vec![Op::Insert; chunk_size],
+                columns,
+                None,
+            ))
+        } else {
+            None
+        })
+    }
+
+    #[try_stream(ok = Watermark, error = StreamExecutorError)]
+    async fn emit_watermarks<'a>(_this: &'a mut ExecutorInner<S>, vars: &'a mut ExecutionVars<S>) {
+        if let Some(last_outputted) = vars.last_outputted_order_key.as_ref() {
+            while let Some(watermark) = vars.pending_watermarks.front() {
+                if &watermark.val > last_outputted {
+                    break;
+                }
+                yield vars.pending_watermarks.pop_front().unwrap();
+            }
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn executor_inner(self) {
+        let OverWindowExecutor {
+            input,
+            inner: mut this,
+        } = self;
+
+        let mut vars = ExecutionVars {
+            partitions: ExecutorCache::new(new_unbounded(this.watermark_epoch.clone())),
+            pending_watermarks: VecDeque::new(),
+            last_outputted_order_key: None,
+            _phantom: PhantomData::<S>,
+        };
+
+        let mut input = input.execute();
+        let barrier = expect_first_barrier(&mut input).await?;
+        this.state_table.init_epoch(barrier.epoch);
+        vars.partitions.update_epoch(barrier.epoch.curr);
+
+        yield Message::Barrier(barrier);
+
+        #[for_await]
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Watermark(watermark) => {
+                    if watermark.col_idx != this.order_key_index {
+                        // Ignore watermark from other columns.
+                        continue;
+                    }
+                    // Note: We may receive many watermarks before we are ready to output rows below
+                    // theme.
+                    vars.pending_watermarks
+                        .push_back(watermark.with_idx(this.output_order_key_index()));
```

Comment:
Since the input executor is a `Sort`, why do we need to buffer watermarks here? Seems that we can produce watermarks

Reply:
Consider `lead(1)`, when we got the row with order key `5`, we may immediately receive watermark `6`, but we cannot emit the result of `5` yet.

Reply:
> Consider `lead(1)`, when we got the row with order key `5`, we may immediately receive watermark `6` after it, but we cannot emit the result of `5` yet.

I mean that the iterator can emit watermark itself when it emits chunks.
If it is too frequent, we can emit a watermark every N rows.

Reply:
Oh you're right.